### ISSUE TITLE
Add threading support to C++ library

### ISSUE THREAD

**Issue by LADSoft** [LADSoft/OrangeC#128]
as part of #85 it was determined that the C++11 library in use does not currently support the THREAD header properly.   Nor does it support mutexes.   The proposal is to copy some support from the C++17 version of the library backwards and leverage the C11 threading library.

**Comment by chuggafan** [comment]
Hmm, I was specifically thinking of creating a separate issue about this to add my context, but it seems that we can leverage the work done on STLPort's latest https://sourceforge.net/p/stlport/code/ci/master/tree/ utilizting the fact that they implemented a mutex implementation, and it does so without using fibersapi, unlike how libcxx did it, which works in our favor since you ALREADY have STLPort in the code, you would just have to update it to this. It would seem to be easier than backporting libcxx, but I see exactly why we would want to use libcxx's ideas for that ourselves.

**Comment by chuggafan** [comment]
Update, I have actually gotten progress on this now due to the fix for #184 happening, but I ran into a constexpr bug where something that (should be?) constexpr isn't in the chrono.cpp implementation I'm taking from libcxx (I'm mixing and maxing c++14 versions, wondrous, but it's working), I've also actually fixed the ```struct timespec``` definition in time.h because it's actually ```time_t tv_sec; long tv_nsec```. If you want to look at it my fork's master branch has all of what I've done on it

**Comment by chuggafan** [comment]
Next in my series of unfortunate events regarding compiler issues: I run into a compiler crash when attempting to build the c++ files from libcxx and your cregex.cpp it seems, I have *LITERALLY* no idea why too because it's giving out a generic windows error code and that's not something I can investigate easily

**Comment by LADSoft** [comment]
I'll snag it off your master in a little bit and see what the problem is.

**Comment by chuggafan** [comment]
Thanks

**Comment by LADSoft** [comment]
i fixed oasm and olib, I had never thought to run a build with debug executables before.   It was a good idea...  occ wasn't generating any crashes for me though.   Although it didn't produce valid output for the C++ files either...  do you need me to look at it?

**Comment by chuggafan** [comment]
Yhea, I'd like for you to take a look at the invalid c++ file output, that's the issue I've run into, it seems to return error code 0xC0000005, and the only way to actually know this is through OMAKE (which says that a negative number was returned)

**Comment by LADSoft** [comment]
I fixed some bad behavior when it was stressed with errors...   now you should be able to see it do something other than crash.   You might want to remove your #defines from __config and fix those errors first though!

**Comment by chuggafan** [comment]
I'm going to work on those errors and see what is a bug with  the actual compiler and differentiate them, I did find a couple of my own bugs thanks to you though! Thanks for the help with this.

**Comment by chuggafan** [comment]
I've fixed the errors I've could, but it seems I've run into two more issues that I'm going to make separate from this and attempt to actually, properly isolate for you. This project seems better than usual at finding bugs, hopefully soon I can actually run occ through csmith and catch some more problems

**Comment by LADSoft** [comment]
hehehe cmith.   I had fun with that one, cleaned up a lot of stuff but got tired of mining after a couple of weeks though.   Maybe you can find some more stuff with it :)

**Comment by chuggafan** [comment]
Yhea, once I can figure it out it should do a lot of good for finding compiler bugs, but that's unrelated to this ;)

**Comment by chuggafan** [comment]
Ok, so progress report: I've made some SIGNIFICANT progress on this (surprisingly enough), I've had to upgrade a HUGE number of headers and other things already, but I've run into unexpected roadblocks that I've had to work around (at least in terms of headers), one such item is once I had to pull in typeinfo from libcxx, as some weird define issue was happening unless I did so, also, doing this broke the headers stdint.h and cstring, for whatever reason, I'm pretty sure I've also run into more than one compiler bug (one minor, one (?) of value) that are in some random place, I'll (hopefully) fix cstring over the next week as well as isolate the other issues that i've run into, but there is significant progress on this task overall.

**Comment by LADSoft** [comment]
geez sounds like this turned into a real project!   I'm glad you are making progress though.    Please do keep me updated on any compiler bugs you find so I can have a chance to fix them :)

**Comment by chuggafan** [comment]
I just checked the libcxx 10 branch, on windows we now have threading support by default if we declare that we have C11 threads. This means that we don't have to "change" anything realistically for building in C++11 threads, the main problem is still going to be compilation however. 

Still, we should probably use the "platform defaults" instead of doing C11 here because we know all the platforms we currently support, are supported, and that the ultramajority of platforms out there either support pthreads or windows threads and our implementation of C11 threads is nothing to write home about.

In short I suggest that despite these additions we use the non-c11 thread system that is offered by libcxx.

**Comment by chuggafan** [comment]
Looking like this will also be solved when #578 gets pulled

**Comment by LADSoft** [comment]
cool :smile:

**Comment by chuggafan** [comment]
With #578 in this is technically done
